### PR TITLE
libCEED - intermediate update prior to v0.7 release

### PR DESCRIPTION
**Pull request by jeremylt**
libCEED has updated our API for `CeedElemRestriction`s to better support mixed finite elements. This change will be part of the libCEED v0.6.1 release.

I have included the updates to the API in this branch. I imagine that there will be other changes required for the libCEED v0.6.1 release, and I am working on figuring out exactly what changes are required.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1436](https://github.com/mfem/mfem/pull/1436) | @jeremylt | @tzanio | @YohannDudouit + @tzanio | 06/30/20 | 06/30/20 | 07/04/20 | |

### CODE REVIEW

**Comment by jedbrown** [comment]
This being a significant API change, the next libCEED release will be v0.7.

**Comment by tzanio** [comment]
Then you want to update also the libCEED requirement in `INSTALL`:

https://github.com/mfem/mfem/blob/master/INSTALL#L643-L648

**Comment by jeremylt** [comment]
@tzanio thank you.

I am not sure if @dmed256 has another OCCA release coming soon, but the new OCCA backend that will be a part of libCEED v0.7 will require OCCA changes beyond v1.0.9, such as https://github.com/libocca/occa/commit/f0b26b77acb3e0b9493faab6d1c380abc94534da.

**Comment by dmed256** [comment]
I have a draft for the next OCCA version I plan to release one the new backend is merged. I'll spend time this weekend updating the backend based on the PR comments. I'm still unsure why there's a output difference between the local runs and Travis CI runs :(

image

File: INSTALL [hunk 1]
```diff
@@ -640,12 +640,12 @@ The specific libraries and their options are:
   Options: OCCA_DIR, OCCA_OPT, OCCA_LIB.
   Versions: OCCA >= 1.0.9.
 
-- libCEED (optional), used when MFEM_USE_CEED = YES. Requires libCEED v0.6
+- libCEED (optional), used when MFEM_USE_CEED = YES. Requires libCEED v0.7
```
Comment (jeremylt) [inline INSTALL:643]: TODO: Update git-hash upon v0.7 release.

**Comment by jeremylt** [comment]
With the release of OCCA v1.0.10, should this PR update MEFM + OCCA as well, or would a concurrent separate PR be preferred?

**Comment by tzanio** [comment]
> With the release of OCCA v1.0.10, should this PR update MEFM + OCCA as well, or would a concurrent separate PR be preferred?

I'll prefer to do it in this PR

File: INSTALL (hunk 1 above)
Comment (tzanio) [reply INSTALL:643]: Actually, ideally this will just say libCEED v0.7. We should get rid of the hash if possible

Comment (jeremylt) [reply INSTALL:643]: Updated.

File: INSTALL [hunk 2]
```diff
@@ -638,14 +638,13 @@ The specific libraries and their options are:
 - OCCA (optional), used when MFEM_USE_OCCA = YES.
   URL: https://libocca.org
   Options: OCCA_DIR, OCCA_OPT, OCCA_LIB.
-  Versions: OCCA >= 1.0.9.
+  Versions: OCCA >= 1.0.10.
```
Comment (v-dobrev) [inline INSTALL:641]: Technically, OCCA >= 1.0.10 is a requirent of libCEED 0.7 -- mfem can still use 1.0.9. Doesn't hurt to update it though. 😃

**Comment by jeremylt** [comment]
This PR is a prerequsite for me adding libCEED's new GPU diagonal assembly capability to MFEM. Would anyone object if I split this PR into

1) libCEED interface update - mergable now
2) libCEED v0.7 - megrable soonish

to get these changes into MFEM sooner so I can start my folloup PR to add libCEED diagonal assembly to MFEM this week?

**Comment by YohannDudouit** [comment]
I think it would be good to merge the libCEED interface update, since currently the two master (MFEM and libCEED) branches don't compile together.

**Comment by tzanio** [comment]
No objections

**Comment by jeremylt** [comment]
In the v0.7 release PR, I'll also add our new interface that will largely replace `CeedVectorSyncArray()`: 
https://github.com/CEED/libCEED/pull/577

File: fem/libceed/ceed.cpp [hunk 3]
```diff
@@ -215,23 +232,32 @@ static void InitCeedTensorBasisAndRestriction(const FiniteElementSpace &fes,
                            grad1d.GetData(), qref1d.GetData(),
                            qweight1d.GetData(), basis);
 
+   CeedInt compstride = fes.GetNDofs();
+   if (fes.GetOrdering()==Ordering::byVDIM)
+   {
+      compstride = 1;
+   }
```
Comment (YohannDudouit) [inline fem/libceed/ceed.cpp:239]:
```suggestion
   CeedInt compstride = fes.GetOrdering()==Ordering::byVDIM ? 1 : fes.GetNDofs();
```

File: fem/libceed/ceed.cpp [hunk 4]
```diff
@@ -97,6 +97,12 @@ static void InitCeedNonTensorBasisAndRestriction(const FiniteElementSpace &fes,
    Vector qweight(Q);
    Vector shape_i(P);
    DenseMatrix grad_i(P, dim);
+
+   CeedInt compstride = fes.GetNDofs();
+   if (fes.GetOrdering()==Ordering::byVDIM)
+   {
+      compstride = 1;
+   }
```
Comment (YohannDudouit) [inline fem/libceed/ceed.cpp:105]:
```suggestion
   CeedInt compstride = fes.GetOrdering()==Ordering::byVDIM ? 1 : fes.GetNDofs();
```

Comment (YohannDudouit) [reply fem/libceed/ceed.cpp:105]: I believe this also fixes a bug no?

**Comment by tzanio** [comment]
LGTM

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**